W-14927671  Pdk ga branch gr (#464)
mulesoft/docs-gateway: pdk/1.0/modules/ROOT/pages/policies-pdk-configure-features-dataweave.adoc

= Using DataWeave Expressions
ifndef::env-site,env-github[]
include::_attributes.adoc[]
endif::[]
:imagesdir: ../assets/images

> 

Reviewer: Does this tutorial have prerequisites? Does it need a "Before You Begin" section? For example, I assume that in order to successfully complete this tutorial you'd need a project that compiles? It may seem obvious that this is the case, but to newbies it might not be obvious.

* `authentication`: Call the `bind_authentication` method with an implementation of the trait `AuthenticationBinding`. PDK provides an implementation of this trait for `AuthenticationData`. To inject authentication information into your policy, see xref:policies-pdk-configure-features-inject-parameters.adoc#authentication[Inject Authentication Inforamaion].
* `payload`: Call the `bind_payload` method with an implementation of the trait `PayloadBinding`. PDK provides an implementation of this trait for `RequestBodyState`, `ResponseBodyState`, and the response of an xref:policies-pdk-configure-features-http-request.adoc[HTTP call].
+
[NOTE]
====
Call only the methods for the bindings defined in the schema definition. For example, if the `attributes` binding is not used, don't call the `bind_attributes` method.

To learn more about defining DataWeave expressions, see xref:policies-pdk-create-schema-definition.adoc#dataweave-expressions[DataWeave Expressions].
====

. After you bind all expression variables, obtain the result by calling the `eval` method.

[[example-evaluator-code]]

> == Example Evaluator Code

Reviewer: Usually in tutorials, "==" level section titles are imperative tasks, something like:
"Evaluate a DataWeave Expression"

But I'm not sure if this document is a tutorial or a reference doc? Is the customer supposed to follow these steps and achieve something? What are they achieving? This kind of info is usually included in a tutorial introduction.

Author: https://github.com/mulesoft/docs-gateway/pull/504/files

> For example, to evaluate the DataWeave expression in the following policy schema definition:

Reviewer: Which expression? "#[vars.myVar]"? I assume so, but I'm not sure at first read. Also what do they have to do to evaluate the expression? Include the expression as the value of `default`?  And what is `default`? Is this explained somewhere? Again, as a newbie I'm not sure what all this means. Will the average customer already know about this stuff?

Author: https://github.com/mulesoft/docs-gateway/pull/504/files

[source,yaml]
----
apiVersion: gateway.mulesoft.com/v1alpha1
kind: Extension
metadata:
 labels:
   title: jwt-auth
   category: Custom
spec:
 extends:
   - name: extension-definition
     namespace: default
 properties:
   tokenExtractor:
     type: string
     format: dataweave
     default: "#[vars.myVar]"
     bindings:
       payload:
         mimeType: text
       attributes: true
       authentication: true
       vars:
         - myVar
 required:
   - tokenExtractor

----

> Use the following Rust snippet:

Reviewer: Is this a step in this tutorial? It looks like a step, but I don't see a series of other steps.

I think I'm confused about if this is a tutorial or a reference doc. Or is it a hybrid? Will the customer also be confused?

[source,rust]
----

> // Copyright 2023 Salesforce, Inc. All rights reserved.

Reviewer: Is this needed? If so, should it be 2024?

   let state = state.into_headers_state().await;
   evaluator.bind_attributes(&HandlerAttributesBinding::new(state.handler(), &stream));

   let state = state.into_body_state().await;
   evaluator.bind_payload(&state);

   if let Ok(value) = evaluator.eval() {
       match value {
           Value::Null => info!("value was null!"),
           Value::Bool(val) => info!("value was Bool: {val}"),
           Value::Number(val) => info!("value was Number: {val}"),
           Value::String(val) => info!("value was String: {val}"),
           Value::Array(val) => info!("value was Array: {val:?}"),
           Value::Object(val) => info!("value was Object: {val:?}"),
       }
   }
}

#[entrypoint]
async fn configure(launcher: Launcher, Configuration(bytes): Configuration) -> Result<()> {
   let config: Config = serde_json::from_slice(&bytes)?;

   launcher
       .launch(on_request(|request, stream, auth| {
           request_filter(request, stream, auth, config.token_extractor.evaluator())
       }))
       .await?;
   Ok(())
}
----

NOTE: This policy uses the `#[attributes.remoteAddress]` variable, which is useful for rate limiting.

> == DefaultBindings

Reviewer: Usually in tutorials, "==" level section titles are imperative tasks, something like:
"Simplify Using the `DefaultBindings` Object"

If you don't require custom bindings in your DataWeave expression, simplify your code by using the `DefaultBindings` object provided by PDK. The `DefaultBindings` object wraps the evaluator and transforms the methods to simplify the resolution of the script.

> `DefaultBindings` requires you to expose the following methods to bind different pieces of data before evaluating the Script:

Reviewer: ```suggestion
`DefaultBindings` requires you to expose the following methods to bind different pieces of data before evaluating the script:
```

[source,rust]
----
fn bind_vars<K: IntoValue>(&mut self, name: &str, value: K)

fn bind_headers(&mut self, binding: &dyn HeadersHandler)

fn bind_payload(&mut self, binding: &dyn PayloadBinding)

fn bind_http_client_response(&mut self, binding: &HttpClientResponse)

fn bind_authentication(&mut self)
----

> For example, to solve the script in <<example-evaluator-code>>, the code needed to inject the `StreamProperties` and `Authentication` parameters. PDK provides these parameters in the `DefaultBindings` object. Given the parameters used in the schema definition from <<example-evaluator-code>>, the expression can be evaluated by using `DefaultBindings` as shown in this Rust snippet:

Reviewer: ```suggestion
For example, to resolve the script in <<example-evaluator-code>>, you must inject the `StreamProperties` and `Authentication` parameters. PDK provides these parameters in the `DefaultBindings` object. Given the parameters used in the schema definition from <<example-evaluator-code>>, the expression can be evaluated by using `DefaultBindings` as shown in this Rust snippet:
```
Suggestion to fix a sentence fragment.

[source,rust]
----

> // Copyright 2023 Salesforce, Inc. All rights reserved.

Reviewer: Need copyright? 2024?

   let mut evaluator = bindings.evaluator(&config.token_extractor);
   evaluator.bind_vars("myVar", "myVal");

   evaluator.bind_authentication();

   let state = state.into_headers_state().await;
   evaluator.bind_headers(state.handler());

   let state = state.into_body_state().await;
   evaluator.bind_payload(&state);

   if let Ok(value) = evaluator.eval() {
       match value {
           Value::Null => info!("value was null!"),
           Value::Bool(val) => info!("value was Bool: {val}"),
           Value::Number(val) => info!("value was Number: {val}"),
           Value::String(val) => info!("value was String: {val}"),
           Value::Array(val) => info!("value was Array: {val:?}"),
           Value::Object(val) => info!("value was Object: {val:?}"),
       }
   }
}

#[entrypoint]
async fn configure(launcher: Launcher, Configuration(bytes): Configuration) -> Result<()> {
   let config: Config = serde_json::from_slice(&bytes)?;

   launcher
       .launch(on_request(|request, bindings| {
           request_filter(request, bindings, &config)
       }))
       .await?;
   Ok(())
}
----

> == Validate the Expression was Solved

Reviewer: I'm not sure what "solved" means. Should it be "resolved"?
```suggestion
== Validate the Expression was Resolved
```

> You may solve some expressions before all values are bound. For example, the expression `#[vars.myVar]` does not require the `attributes`, `payload`, no `authentication` to be solved. 

Reviewer: ```suggestion
You can resolve some expressions before all values are bound. For example, the expression `#[vars.myVar]` does not require the `attributes`, `payload`, no `authentication` to be resolved. 
```
"May" implies we're giving the customer permission, which sounds weird. "Can" implies something they can do.

Also, I'm confused about "solve" vs. "resolve".

> After one of the binding methods is called, you can confirm the expression was completely solved by calling the method:

Reviewer: ```suggestion
After one of the binding methods is called, you can confirm the expression was completely resolved by calling the method:
```

[source,rust]
----

> evaluator.is_ready() method.

Reviewer: ```suggestion
evaluator.is_ready()
```

= Using DataWeave Expressions
ifndef::env-site,env-github[]
include::_attributes.adoc[]
endif::[]
:imagesdir: ../assets/images

NOTE: To view an example policy project that uses DataWeave expressions, see https://github.com/mulesoft/pdk-custom-policy-examples/blob/main/query/README.md[Query Policy Example^].

> If input DataWeave expression parameter is valid, the policy parses the expression and transforms it into a `pdk::script::Script` expression. To obtain the result of the script:

Reviewer 2: ```suggestion
If the input DataWeave expression parameter is valid, the policy parses the expression and transforms it into a `pdk::script::Script` expression. To obtain the result of the script:
```

. Create an evaluator.

> . Call the binding function for each parameter defined in schema definition:

Reviewer 2: ```suggestion
. Call the binding function for each parameter defined in the schema definition:
```

+
* `vars`: Call the `bind_vars` method for each of the policy's `vars`. After all of `vars` are bound, the evaluator tries to solve the expression. If the expression contains `vars` other than the ones defined in the schema, they resolve to `null`.

> * `attributes`: Call the `bind_attributes` methods with an implementation of the `AttributesBinding` trait. PDK provides a implementation of this trait by instantiating a new `HandlerAttributesBinding` for each `RequestHeaderState`, `RequestHeaderState`, and response of an xref:policies-pdk-configure-features-http-request.adoc[HTTP call].

Reviewer 2: ```suggestion
* `attributes`: Call the `bind_attributes` method with an implementation of the `AttributesBinding` trait. PDK provides an implementation of this trait by instantiating a new `HandlerAttributesBinding` for each `RequestHeaderState`, `RequestHeaderState`, and response of an xref:policies-pdk-configure-features-http-request.adoc[HTTP call].
```

> * `authentication`: Call the method `bind_authentication` with an implementation of the trait `AuthenticationBinding`. PDK provides an implementation of this trait for `AuthenticationData`. To inject authentication information into your policy, see xref:policies-pdk-configure-features-inject-parameters.adoc#authentication[Inject Authentication Inforamaion].

Reviewer 2: ```suggestion
* `authentication`: Call the `bind_authentication` method with an implementation of the trait `AuthenticationBinding`. PDK provides an implementation of this trait for `AuthenticationData`. To inject authentication information into your policy, see xref:policies-pdk-configure-features-inject-parameters.adoc#authentication[Inject Authentication Inforamaion].
```

* `payload`: Call the `bind_payload` method with an implementation of the trait `PayloadBinding`. PDK provides an implementation of this trait for `RequestBodyState`, `ResponseBodyState`, and the response of an xref:policies-pdk-configure-features-http-request.adoc[HTTP call].
+
[NOTE]
====

> You only need to call the methods for the bindings that are defined in the schema definition. For example, if the `attributes` binding is not used, you don't need to call the `bind_attributes` method.

Reviewer 2: ```suggestion
Call only the methods for the bindings defined in the schema definition. For example, if the `attributes` binding is not used, don't call the `bind_attributes` method.
```

To learn more about defining DataWeave expressions, see xref:policies-pdk-create-schema-definition.adoc#dataweave-expressions[DataWeave Expressions].
====

> . After you bind all expression variables, you can obtain the result by calling the `eval` method.

Reviewer 2: ```suggestion
. After you bind all expression variables, obtain the result by calling the `eval` method.
```

[[example-evaluator-code]]
== Example Evaluator Code

> For example, the DataWeave expression in the following policy schema definition:

Reviewer 2: ```suggestion
For example, to evaluate the DataWeave expression in the following policy schema definition:
```

[source,yaml]
----
apiVersion: gateway.mulesoft.com/v1alpha1
kind: Extension
metadata:
 labels:
   title: jwt-auth
   category: Custom
spec:
 extends:
   - name: extension-definition
     namespace: default
 properties:
   tokenExtractor:
     type: string
     format: dataweave
     default: "#[vars.myVar]"
     bindings:
       payload:
         mimeType: text
       attributes: true
       authentication: true
       vars:
         - myVar
 required:
   - tokenExtractor

----

> Can be evaluated with this Rust snippet:

Reviewer 2: ```suggestion
Use the following Rust snippet:
```

   let state = state.into_headers_state().await;
   evaluator.bind_attributes(&HandlerAttributesBinding::new(state.handler(), &stream));

   let state = state.into_body_state().await;
   evaluator.bind_payload(&state);

   if let Ok(value) = evaluator.eval() {
       match value {
           Value::Null => info!("value was null!"),
           Value::Bool(val) => info!("value was Bool: {val}"),
           Value::Number(val) => info!("value was Number: {val}"),
           Value::String(val) => info!("value was String: {val}"),
           Value::Array(val) => info!("value was Array: {val:?}"),
           Value::Object(val) => info!("value was Object: {val:?}"),
       }
   }
}

#[entrypoint]
async fn configure(launcher: Launcher, Configuration(bytes): Configuration) -> Result<()> {
   let config: Config = serde_json::from_slice(&bytes)?;

   launcher
       .launch(on_request(|request, stream, auth| {
           request_filter(request, stream, auth, config.token_extractor.evaluator())
       }))
       .await?;
   Ok(())
}
----

NOTE: This policy uses the `#[attributes.remoteAddress]` variable, which is useful for rate limiting.

== DefaultBindings

> If you don't require custom bindings in your DataWeave expression, you can simplify your code by using the `DefaultBindings` object provided by PDK. The `DefaultBindings` object that wraps the evaluator and transforms the methods to simplify the resolution of the script.

Reviewer 2: ```suggestion
If you don't require custom bindings in your DataWeave expression, simplify your code by using the `DefaultBindings` object provided by PDK. The `DefaultBindings` object wraps the evaluator and transforms the methods to simplify the resolution of the script.
```

> For example, to solve the script in <<example-evaluator-code>>, the code needed to inject the `StreamProperties` and `Authentication` parameters. PDK provides these parameters in the `DefaultBindings` object. 

Reviewer 2: Unclear sentence, please reword.
